master/internal/rm/kubernetesrm/pods.go

```diff
@@ -864,6 +908,157 @@ func (p *pods) podStatusCallback(s *actor.System, event watch.Event) {
+	} else if err != nil {
+		return nil, fmt.Errorf(
+			"enabling node %s by removing the Determined no schedule label: %w", nodeName, err)
+	}
+	p.syslog.Infof("node %s enabled by an user", nodeName)
+
+	n, ok := p.summarizeClusterByNodes(ctx)[nodeName]
+	if !ok {
+		return nil, fmt.Errorf("node %s enabled without error, error getting node summary", nodeName)
+	}
+	n.Enabled = true
+	n.Draining = false
+	for slotKey := range n.Slots {
+		s := n.Slots[slotKey]
+		s.Enabled = n.Enabled
+		s.Draining = n.Draining
+		n.Slots[slotKey] = s
+	}
+
+	return &apiv1.EnableAgentResponse{
+		Agent: n.ToProto(),
+	}, nil
+}
+
+func (p *pods) disableNode(
+	ctx *actor.Context, nodeName string, shouldDrain bool,
+) (*apiv1.DisableAgentResponse, error) {
+	labelValue := noExecuteNodeLabelValue
+	if shouldDrain {
+		labelValue = noScheduleNodeLabelValue
+	}
+
+	patchStruct := metaV1.ObjectMeta{
+		Labels: map[string]string{clusterIDNodeLabel(): labelValue},
+	}
+	patch, err := json.Marshal(map[string]any{"metadata": patchStruct})
+	if err != nil {
+		return nil, fmt.Errorf("marshaling JSON patch %v: %s", patchStruct, err)
+	}
+
+	_, err = p.clientSet.CoreV1().Nodes().
+		Patch(context.TODO(), nodeName, types.StrategicMergePatchType, patch, metaV1.PatchOptions{})
+	if k8error.IsForbidden(err) {
+		return nil, fmt.Errorf("the Determined master Kubernetes service account " +
+			"is missing permissions to patch nodes. " +
+			"Enabling or disabling nodes requires this permission, " +
+			"however Determined will otherwise still function correctly without " +
+			"these Kubernetes permissions")
+	} else if err != nil {
+		return nil, fmt.Errorf(
+			"disabling node %s by adding the Determined no schedule label: %w", nodeName, err)
+	}
+	p.syslog.Infof("node %s disabled by an user", nodeName)
+
+	if !shouldDrain { // See note in spec.go about how we could remove killing all pods here.
+		if err := p.releaseAllocationsOnDisabledNode(ctx, nodeName); err != nil {
+			return nil, fmt.Errorf(
+				"node disabled without error, error killing existing pod on node: %w", err)
+		}
+	}
+
+	n, ok := p.summarizeClusterByNodes(ctx)[nodeName]
+	if !ok {
+		return nil, fmt.Errorf("node %s disabled without error, error getting node summary", nodeName)
+	}
+	n.Enabled = false
+	n.Draining = shouldDrain
+	for slotKey := range n.Slots {
+		s := n.Slots[slotKey]
+		s.Enabled = n.Enabled
+		s.Draining = n.Draining
+		n.Slots[slotKey] = s
+	}
+
+	return &apiv1.DisableAgentResponse{
+		Agent: n.ToProto(),
+	}, nil
+}
+
+func (p *pods) releaseAllocationsOnDisabledNode(ctx *actor.Context, nodeName string) error {
+	listOptions := metaV1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s", determinedLabel),
+		FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodeName),
+	}
+	pods, err := p.listPodsInAllNamespaces(context.TODO(), listOptions)
+	if err != nil {
+		return fmt.Errorf("listing pods on node %s: %w", nodeName, err)
+	}
+
+	allocIDs := make(map[model.AllocationID]bool)
```

Comment:
better name, and then we don't really need the comment on 1048
```suggestion
	notifiedAllocations := make(map[model.AllocationID]bool)
```